### ISSUE TITLE
ZStream: combinator similar to akka-streams takeWithin

### ISSUE THREAD

**Issue by juliano** [zio/zio#3465]
Hello all,

[After asking in the channel](https://discordapp.com/channels/629491597070827530/629497941719121960/703911274484072488) about something similar to [`akka-streams takeWithin`](https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/takeWithin.html), as suggested by @iravid, it would be nice to have a combinator for that.

Thanks!

**Comment by luis3m** [comment]
```scala
def takeWithin(duration: Duration): ZStream[R with Clock, E, O] =
  ZStream {
    for {
      chunks <- self.process
      time   <- clock.nanoTime.toManaged_
      expire =  time + duration.toNanos
      pull   =  for {
         chunk <- chunks
         now   <- clock.nanoTime
         pull  <- if (now > expire) Pull.end else Pull.emit(chunk)
      } yield pull
    } yield pull
  }
```

I guess it's as simple as this, @iravid isn't it?

**Comment by iravid** [comment]
@luis3m Unfortunately not, because if you're pulling from, say, a socket - then `chunks` could semantically block for a longer period than `duration`. It might actually be just `stream.interruptWhen(ZIO.sleep(duration))` 🧐

I'd call it `ZStream#interruptAfter` in this case.

We can also do a variant `ZStream#haltAfter` that doesn't interrupt long-running pulls.

**Comment by luis3m** [comment]
@iravid right, it's like adding a `timeout` as long as `durarion`  to avoid unnecessary block. Anyway, seems like `interruptWhen` will do it as you suggested which makes it even easier. I can submit a PR for both `interruptAfter` and `halfAlfter`.

**Comment by iravid** [comment]
@luis3m That'd be great!

**Comment by luis3m** [comment]
@juliano it's merged now and part of `RC19`

**Comment by iravid** [comment]
Resolved in #3534
